src/backend/cpu/kernel/canny.hpp

```diff
@@ -0,0 +1,237 @@
+                                }
+                            } else {
+                                if ((dX[offset]+dY[offset]) >= 0) {
+                                    // East of North East
+                                    alpha = (float)-dY[offset]/dX[offset];
+                                    mag1  = (1-alpha)*mag[offset+1] + alpha*mag[offset-dims[0]+1];
+                                    mag2  = (1-alpha)*mag[offset-1] + alpha*mag[offset+dims[0]-1];
+                                } else {
+                                    // North of North East
+                                    alpha = (float)dX[offset]/-dY[offset];
+                                    mag1  = (1-alpha)*mag[offset-dims[0]] + alpha*mag[offset-dims[0]+1];
+                                    mag2  = (1-alpha)*mag[offset+dims[0]] + alpha*mag[offset+dims[0]-1];
+                                }
+                            }
+                        } else {
+                            if (dY[offset]>=0) {
+                                if ((dX[offset]+dY[offset]) >= 0) {
+                                    // South of South West
+                                    alpha = (float)-dX[offset]/dY[offset];
+                                    mag1  = (1-alpha)*mag[offset+dims[0]] + alpha*mag[offset+dims[0]-1];
+                                    mag2  = (1-alpha)*mag[offset-dims[0]] + alpha*mag[offset-dims[0]+1];
+                                } else {
+                                    // West of South West
+                                    alpha = (float)dY[offset]/-dX[offset];
+                                    mag1  = (1-alpha)*mag[offset-1] + alpha*mag[offset+dims[0]-1];
+                                    mag2  = (1-alpha)*mag[offset+1] + alpha*mag[offset-dims[0]+1];
+                                }
+                            } else {
+                                if ((-dX[offset]+dY[offset]) >= 0) {
+                                    // West of North West
+                                    alpha = (float)-dY[offset]/dX[offset];
+                                    mag1  = (1-alpha)*mag[offset-1] + alpha*mag[offset-dims[0]-1];
+                                    mag2  = (1-alpha)*mag[offset+1] + alpha*mag[offset+dims[0]+1];
+                                } else {
+                                    // North of North West
+                                    alpha = (float)dX[offset]/-dY[offset];
+                                    mag1  = (1-alpha)*mag[offset-dims[0]] + alpha*mag[offset-dims[0]-1];
+                                    mag2  = (1-alpha)*mag[offset+dims[0]] + alpha*mag[offset+dims[0]+1];
+                                }
+                            }
+                        }
+
+                        if (mag[offset]>mag1 && mag[offset]>mag2) {
+                            out[offset] = mag[offset] > 255 ? EDGE : (unsigned char)mag[offset];
```

Comment:
I am not sure, may be i have to revisit the algorithm details once again. The output array is of `uchar` type, therefore i didn't think it would be a problem earlier.